lockfile/src/cyclonedx.rs

```diff
@@ -0,0 +1,313 @@
+use std::ffi::OsStr;
+use std::path::Path;
+use std::str::FromStr;
+
+use anyhow::anyhow;
+use phylum_types::types::package::PackageType;
+use purl::GenericPurl;
+use serde::Deserialize;
+use thiserror::Error;
+
+use crate::{Package, PackageVersion, Parse, ThirdPartyVersion};
+
+// Define a custom error for unknown ecosystems.
+#[derive(Error, Debug)]
+#[error("Could not determine ecosystem")]
+struct UnknownEcosystem;
+
+// Define the generic trait for components.
+trait Component {
+    fn component_type(&self) -> &str;
+    fn name(&self) -> &str;
+    fn version(&self) -> &str;
+    fn scope(&self) -> Option<&str>;
+    fn purl(&self) -> Option<&str>;
+    fn components(&self) -> Option<&[Self]>
+    where
+        Self: Sized;
+}
+
+// CycloneDX BOM
+#[derive(Clone, Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct Bom<T> {
+    components: Option<T>,
+}
+
+// Struct for wrapping a list of components from XML
+#[derive(Clone, Debug, Deserialize)]
+struct Components<T> {
+    #[serde(rename = "component")]
+    components: Vec<T>,
+}
+
+// Represents a single XML component.
+#[derive(Clone, Debug, Deserialize)]
+struct XmlComponent {
+    #[serde(rename = "type")]
+    component_type: String,
+    name: String,
+    version: String,
+    scope: Option<String>,
+    purl: Option<String>,
+    components: Option<Components<XmlComponent>>,
```

Comment:
Do I see it correctly that the only reason for having separate structs between XML and JSON is this difference in array handling of XML?

Reply:
Yes, something to do with the attribute. Is there's a better way to handle that?

Seems to be something to do with the attribute.
```<component type="library"```

Reply:
Honestly I'm not **entirely** sure. It would be possible to share the fields that are in common I suppose.